test/shared/src/main/scala/zio/test/Fun.scala

```diff
@@ -0,0 +1,40 @@
+package zio.test
+
+import zio.ZIO
+
+/**
+ * A `Fun[A, B]` is a referentially transparent version of a potentially
+ * effectual function from `A` to `B`. Each invocation of the function will be
+ * memoized so the function is guaranteed to return the same value for any
+ * given input. The function should not involve asynchronous effects.
+ */
+final case class Fun[-A, +B] private (private val f: A => B, private val hash: A => Int) extends (A => B) {
```

Comment:
We should make this and the companion object package private, so that it doesn't appear in the Scaladoc (unless it needs to be public?).

Reply:
We can make it private. I thought a next step could be exploring building this out into something along the lines of shrinkable and showable functions in QuickCheck, though even then I'm not sure we would need to expose it and we could always do it later.